### PR TITLE
refactor: Add RF and Tokens to Loki ring Cfg and allow overwriting docs.

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:

This PR adds two new fields to the loki ring config: num tokens and replication factor.
The default values for the replication factor and the number of tokens varies for different components. On top of that, some components doesn't use those configs at all (they hardcode them).

To overcome that, this PR supports overwriting the default values and docs for any of the config fields in the loki ring. We do this via a new optional argument to the `RegisterFlagsWithPrefix` method with a list of flags to skip defining. By passing this arg the ring itself won't register the flags for those fields and now it's up to the caller to define them with different defaults and docs.

This opens the door to having the loki ring config as a wrapper of the dskit ring lifecycler config and:
-  Documenting the values of the ring config that won't be configurable (i.e. will be ignored) for each loki component independently.
- Set different default values for each ring config field.
- Add more in-depth docs to any of the ring config fields whenever it's convennient. E.g. the bloom compactor defines docs for the num tokens explaining the effects of increasing it.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/stores/shipper/indexshipper/indexgateway/config.go [hunk 1]
```diff
@@ -76,11 +62,31 @@ type Config struct {
...
+	//
+	// Whenever the store queries the ring key-value store for the Index Gateway instance responsible for tenant X,
+	// multiple Index Gateway instances are expected to be returned as Index Gateway might be busy/locked for specific
+	// reasons (this is assured by the spikey behavior of Index Gateway latencies).
+	f.IntVar(&cfg.Ring.ReplicationFactor, "replication-factor", ReplicationFactor, "Deprecated: How many index gateway instances are assigned to each tenant. Use -index-gateway.shard-size instead. The shard size is also a per-tenant setting.")
+}
+
+func (cfg *Config) Validate() error {
+	if cfg.Ring.NumTokens != NumTokens {
+		return errors.New("Num tokens must not be changed as it will not surge effect")
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/storage/stores/shipper/indexshipper/indexgateway/config.go:89]: "won't take effect"

File: pkg/bloomcompactor/config.go [hunk 2]
```diff
@@ -48,6 +47,17 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
+	// Ring
+	skipFlags := []string{
+		"bloom-compactor.ring.num-tokens",
+		"bloom-compactor.ring.replication-factor",
+	}
+	cfg.Ring.RegisterFlagsWithPrefix("bloom-compactor.", "collectors/", f, skipFlags...)
+	// Overrides
+	f.IntVar(&cfg.Ring.NumTokens, "bloom-compactor.ring.num-tokens", 10, "Number of tokens to use in the ring. The bigger the number of tokens, the more fingerprint ranges the compactor will own, but the smaller these ranges will be. Bigger number of tokens will result in more and smaller metas and blocks.")
+	// Ignored
+	f.IntVar(&cfg.Ring.ReplicationFactor, "bloom-compactor.ring.replication-factor", 1, "IGNORED: Replication factor is fixed to 1")
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/bloomcompactor/config.go:60]: Added check in `Validate`

File: pkg/loki/config_wrapper.go [hunk 3]
```diff
@@ -314,6 +314,8 @@ func applyConfigToRings(r, defaults *ConfigWrapper, rc lokiring.RingConfig, merg
 		r.BloomCompactor.Ring.InstanceZone = rc.InstanceZone
 		r.BloomCompactor.Ring.ZoneAwarenessEnabled = rc.ZoneAwarenessEnabled
 		r.BloomCompactor.Ring.KVStore = rc.KVStore
+		r.BloomCompactor.Ring.NumTokens = rc.NumTokens
+		r.BloomCompactor.Ring.ReplicationFactor = rc.ReplicationFactor
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/loki/config_wrapper.go:318]: Removed

File: pkg/loki/config_wrapper.go [hunk 4]
```diff
@@ -327,6 +329,8 @@ func applyConfigToRings(r, defaults *ConfigWrapper, rc lokiring.RingConfig, merg
 		r.BloomGateway.Ring.InstanceZone = rc.InstanceZone
 		r.BloomGateway.Ring.ZoneAwarenessEnabled = rc.ZoneAwarenessEnabled
 		r.BloomGateway.Ring.KVStore = rc.KVStore
+		r.BloomGateway.Ring.NumTokens = rc.NumTokens
+		r.BloomGateway.Ring.ReplicationFactor = rc.ReplicationFactor
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/loki/config_wrapper.go:333]: Moved

**Review by cyriltovena — APPROVED** [review]
LGTM

Not an expert on this but I don't see anything wrong.

**Review by owen-d — APPROVED** [review]

**Pull request merged**